broadinstitute/single_cell_portal_core
# Expanding keyword search to study metadata (SCP-4275)

Currently, keyword searches in SCP only look at study titles/descriptions, and also author information.  This update extends that functionality to now look through stored search facet entries to determine if a provided keyword matches a given facet filter.  Upon matching, this will trigger a metadata search in MongoDB (faster than BigQuery, and contains the same data) to find all possible hits.  These additional studies are then rolled into the normal keyword-based matches, and a combined weight of text + metadata hits is calculated to appropriately weight results.

For instance, searching for terms like `blood` and `tuberculosis` will hit on the `organ` and `disease` facets, and then append any studies that have those values present in their metadata files.  Furthermore, since this is a keyword rather than faceted search, partial matches (whole words, not substrings) are also included.  This means that `tuberculosis` will match on filters such as `esophageal tuberculosis`, `intestinal tuberculosis`, and `oral tuberculosis`.

Screenshot showing new keyword metadata matches of the above scenario:
![Screen Shot 2022-04-22 at 3 34 44 PM](https://user-images.githubusercontent.com/729968/164782368-e1b0dd1b-d293-4cb6-9676-25bc0b8af0ef.png)

In addition, this update refactors some of the logic in `Api::V1::SearchController` into the `StudySearchService`, with the intent being better test coverage.  As such, new test suites have been added for `StudySearchResultsObjects` and `StudySearchService`.

MANUAL TESTING
1. Pull this branch and boot as normal
2. Search for `blood tuberculosis` in the text search box
3. Confirm that the synthetic studies `Tuberculosis subtypes in human male, blood study with arrays` and `Human HIV and Tuberculosis, blood study` both show up first, and have both highlighted text matches and metadata badges

This PR satisfies SCP-4275.
